### ISSUE TITLE
`radeco` utility request/specification

### ISSUE THREAD

**Issue by XVilka** [radare/radeco-lib#13]
Radeco
======

The radare decompiler tool, (namedly 'radeco') should be able to run from r2 or from the shell, accepting different inputs and dumping different outputs.

we can make radeco r2pipe friendly and interact with r2 using the pipes. this piece is already done, but not used at all. so i would suggest making the 'radeco' executable that allows to run like this:
    
    $ radeco -h
    Usage: radeco [-flags] [...]
    -h           show this help
    -o [type]    specify output type (ssa, )
    -i [type]    specify input type
    -r [r2pipe]  specify how to connect to r2
    -O           output all intermediate representations to files (.asm .esil .ssa. ...)
    Types: (listedn bottom-top)
     - asm   : assembly/hexpairs
     - esil
     - ssa
     - dot
     - ... magic ...
     - C code

The -r flag can be used to tell the r2pipe connection how to talk to r2 using http, pipes, etc. in order to retrieve all the type information, function signatures, function names, symbols, imports, variable names, etc. By default, if no -r is specified it should check the environment variable R2PIPE_IN and R2PIPE_OUT and connect using pipe in that case.

We should be able to run this tool from commandline or inside r2 like this:
    
    $ radeco -i asm -o ssa disasm.txt > disasm-ssa.txt

And from inside the r2 shell:
    
    > #!pipe radeco -o c   # decompile current function

As long as we have a chain of transformations we should define that pipeline in code and be able to serialize that info from memory data structures to something human readable. This will be good for the testsuite and for debugging issues on every internal decompilation step.

Bear in mind that radeco will get info from r2... but it will also export info to r2. this is. radeco is able to identify basic blocks, local variables, loops, etc.

-> we shold focus on having this tool done asap, it will help to understand the whole problem better, and everyone should be able to test it and use it from r2. The tool can be then rewritten as an RCore command, but this is out of the scope of the current problem, because decompilation needs to happen first.

The next step would be to write a PseudoWriter module that takes the graph and prints C-like code, the graph representation must contain all the information needed to generate this output.

**Comment by mewmew** [comment]
> As long as we have a chain of transformations we should define that pipeline in code and be able to serialize that info from memory data structures to something human readable. This will be good for the testsuite and for debugging issues on every internal decompilation step.

There is no +1 on GitHub issues, but I'd like echo this idea. I wholeheartedly believe in the powerful notion of having tools operate on well-defined input and output, and having these tools emit data from intermediate steps in human readable format (or a format which is isomorphic to a human readable format). This approach facilitates debugging and the implementation of test cases, it gives great insight into the inner workings of the pipeline, and most importantly, it enables multiple implementations of each component within the pipeline. Each implementation with their respective benefits and trade offs.

I know radare has always followed the Unix philosophy. I'm just happy to see radeco following the same principles.

**Comment by sushant94** [comment]
This issue was moved to radare/radeco#2
